### PR TITLE
feat(GPS): allow send attributes in Google PubSub message.

### PR DESCRIPTION
**Pull request by p-pichet**
Allow to send PubSub attributes

### CODE REVIEW

**Comment by p-pichet** [comment]
@makasim Can someone look at this PR please ?

**Comment by p-pichet** [comment]
Hello @dgafka. 
Can you review this PR or tell me who can ?

**Comment by dgafka** [comment]
Hello @p-pichet 
I am not maintainer of this package, @makasim is ;)

**Comment by p-pichet** [comment]
as i saw you approve some requests and i don't have any responses, i ask you but i understand.

**Comment by p-pichet** [comment]
@makasim

**Comment by makasim** [comment]
ci pipeline does not trigger for this PR. I dont know why. I cannot merge PR without it being tested on CI succesfully.

**Comment by p-pichet** [comment]
> ci pipeline does not trigger for this PR. I dont know why. I cannot merge PR without it being tested on CI succesfully.

i see why the CI is not trigger. 
You are agree with the modification ?

**Comment by p-pichet** [comment]
I don't understand how to run CI.

**Comment by p-pichet** [comment]
Hi @makasim, can you run CI actions, please?

**Comment by makasim** [comment]
I cannot, I would do it if it were possible

**Comment by p-pichet** [comment]
I hesitate if it was a trigger message. 
Do you know the condition to enable the CI ? 
Is it because my repository is not correctly setup ?

**Comment by makasim** [comment]
CI has to be triggered on any MR by default without any additional actions from any side. There is one exception where a maintainer has to approve CI but there is none.

**Comment by p-pichet** [comment]
@makasim i push a fix, can you run the CI please.

**Comment by p-pichet** [comment]
@makasim i push a change, can you run the CI please.

**Comment by p-pichet** [comment]
@makasim i push a change, can you run the CI please.

**Comment by p-pichet** [comment]
@makasim I push a change, can you run the CI please.
For the version 8.x of PHP, we got issue on the SQS, but I don't touch it. Who can have a look at it ?

**Comment by p-pichet** [comment]
@makasim We found that the 404 error come from an update of aws/aws-sdk-php. 
Who is in charge of the maintenance of external lib ?

**Comment by p-pichet** [comment]
@makasim I push a change, can you run the CI please.

**Comment by p-pichet** [comment]
@makasim it's seem the image use to run CI not found the docker compose command.

**Comment by p-pichet** [comment]
@makasim I push a change, can you run the CI please.

**Comment by p-pichet** [comment]
@makasim  i push again 
please run the ci

**Comment by p-pichet** [comment]
@makasim can you run the CI please ?

**Comment by Steveb-p** [comment]
You can ping me instead :sweat_smile: 

I've ran most of your requests recently :)

**Comment by p-pichet** [comment]
thanks @Steveb-p . 
Have you see my other messages ?

**Comment by Steveb-p** [comment]
> thanks @Steveb-p . Have you see my other messages ?

Yeah, but I'm mostly in the same situation as you are. I was mostly involved in rdkafka itself when I was working on queues in my old job, and can't really help with those custom docker images :confounded:

**Comment by p-pichet** [comment]
@Steveb-p  can you run the CI please

**Comment by p-pichet** [comment]
@Steveb-p i try something can you run the CI ?

**Comment by p-pichet** [comment]
@Steveb-p and @makasim, the CI fail on the SQS and SNS. 
We don't have strong knowledge to understand what happen. 
do you have a name to someone can help use ?
Maybe @ASKozienko as you work on the failing test.

**Comment by p-pichet** [comment]
@Steveb-p can you run the CI please ?

**Comment by p-pichet** [comment]
Thnaks @Steveb-p 
The CI is green. 
Who can merge it ?

**Comment by Steveb-p** [comment]
> Thanks @Steveb-p The CI is green. Who can merge it ?

Let me review the code after work, just in case.

File: pkg/gps/GpsMessage.php [hunk 1]
```diff
@@ -169,8 +176,13 @@ public function getNativeMessage(): ?GoogleMessage
         return $this->nativeMessage;
     }
 
-    public function setNativeMessage(GoogleMessage $message = null): void
+    public function setNativeMessage(?GoogleMessage $message = null): void
     {
         $this->nativeMessage = $message;
     }
+
+    public function getAttributes(): array
```
Comment (Steveb-p) [inline pkg/gps/GpsMessage.php:184]: I don't see a `setAttributes()` setter method. I'd assume that something someone might want to do.

In your current code the only way to manipulate the attributes is to pass them via constructor, as headers (`$headers['attributes']`). They cannot be changed afterwards, and they are only "moved" to `attributes` property (and `unset` from `$headers` array) in constructor.

I'd suggest removing the changes from the constructor, and expect users to pass attributes using `setAttributes()` method only. This way there will be no "hidden" functionality in headers and there will be no discrepancy between `setHeaders()` and constructor.

WDYT?

Comment (p-pichet) [reply pkg/gps/GpsMessage.php:184]: I do like this to make it usable without completion on the lib we use ( https://github.com/sroze/messenger-enqueue-transport).

File: composer.json [hunk 2]
```diff
@@ -30,7 +30,7 @@
     "doctrine/persistence": "^2.0|^3.0",
     "mongodb/mongodb": "^1.2",
     "pda/pheanstalk": "^3.1",
-    "aws/aws-sdk-php": "^3.155",
+    "aws/aws-sdk-php": "^3.319",
```
[the remark this answers is not in the document]
Comment (p-pichet) [reply composer.json]: @Steveb-p  i set the lowest version that success. 
can you run CI please ?

File: pkg/gps/GpsMessage.php (hunk 1 above)
Comment (p-pichet) [reply pkg/gps/GpsMessage.php:184]: @Steveb-p i check and don't understand what is the `$properties` in `GPSMessage`. 
What this should be in ?

Comment (Steveb-p) [reply pkg/gps/GpsMessage.php:184]: I wasn't using that library, I don't know what it represents too :sweat_smile: 

Anyway, I understand your reasoning - why you put attributes as a property of headers array. You're trying to make it interoperable between packages.

To make it consistent I'd suggest:
1. Making `setHeaders` work in the exact same way, i.e. unpack the `$headers['attributes']` into a dedicated property. You could actually just call this method from the constructor, so you won't duplicate the code/behavior.
2. Add the explicit `setAttributes()` method regardless, for those that don't want the interoperability aspect and need that explicit call.

Comment (p-pichet) [reply pkg/gps/GpsMessage.php:188]: I change to add attributes instead of using a key in header

**Comment by p-pichet** [comment]
@Steveb-p  can you run the CI and take a new look on the code please ?

File: pkg/sns/composer.json [hunk 3]
```diff
@@ -9,7 +9,7 @@
         "php": "^7.4|^8.0",
         "queue-interop/queue-interop": "^0.8",
         "enqueue/dsn": "^0.10",
-        "aws/aws-sdk-php": "~3.155"
+        "aws/aws-sdk-php": "~3.319"
```
Comment (Steveb-p) [inline pkg/sns/composer.json:12]:
```suggestion
        "aws/aws-sdk-php": "~3.155"
```

**Comment by p-pichet** [comment]
@Steveb-p, I push the missing code style. 
Can you run the CI please ?
And as you validate the PR, what I need to do for its merge ?

File: pkg/sqs/composer.json [hunk 4]
```diff
@@ -9,7 +9,7 @@
         "php": "^7.4|^8.0",
         "queue-interop/queue-interop": "^0.8",
         "enqueue/dsn": "^0.10",
-        "aws/aws-sdk-php": "~3.155"
+        "aws/aws-sdk-php": "~3.319"
```
Comment (Steveb-p) [inline pkg/sqs/composer.json:12]:
```suggestion
        "aws/aws-sdk-php": "~3.155"
```

File: pkg/sns/composer.json (hunk 3 above)
Comment (p-pichet) [reply pkg/sns/composer.json]: I change to use he same version in the global composer

File: pkg/sqs/composer.json (hunk 4 above)
Comment (p-pichet) [reply pkg/sqs/composer.json]: same here.

**Comment by p-pichet** [comment]
Forgot the part to get the attributes on receive a message.

File: bin/changelog [hunk 5]
```diff
@@ -8,6 +8,6 @@ then
     exit 1
 fi
 
-docker-compose run -e CHANGELOG_GITHUB_TOKEN=${CHANGELOG_GITHUB_TOKEN:-""} --workdir="/mqdev" --rm generate-changelog github_changelog_generator --future-release "$1" --no-issues --unreleased-only --output "CHANGELOG_FUTURE.md"
+docker compose run -e CHANGELOG_GITHUB_TOKEN=${CHANGELOG_GITHUB_TOKEN:-""} --workdir="/mqdev" --rm generate-changelog github_changelog_generator --future-release "$1" --no-issues --unreleased-only --output "CHANGELOG_FUTURE.md"
 
-#git add CHANGELOG.md && git commit -m "Release $1" -S && git push origin "$CURRENT_BRANCH"
\ No newline at end of file
+#git add CHANGELOG.md && git commit -m "Release $1" -S && git push origin "$CURRENT_BRANCH"
```
Comment (makasim) [inline bin/changelog:13]: uncomment please, I need it while releasing new tag.

File: docker-compose.yml [hunk 6]
```diff
@@ -127,13 +127,13 @@ services:
       - '9090:9090'
 
   localstack:
-    image: 'localstack/localstack:0.8.10'
+    image: 'localstack/localstack'
```
Comment (makasim) [inline docker-compose.yml:130]: set a fixed version please

**Comment by makasim** [comment]
Few comments from me, Almost there.

**Comment by makasim** [comment]
@p-pichet big thank you for your tireless work. Enqueue's CI is green and it not only provides the feature you need but unblocks a lot of other stuck MRs. 

Kudos to you, much appreciated!

**Comment by p-pichet** [comment]
> @p-pichet big thank you for your tireless work. Enqueue's CI is green and it not only provides the feature you need but unblocks a lot of other stuck MRs.
> 
> Kudos to you, much appreciated!

Thanks a lot. 
When do you think the release will be created ?

**Comment by makasim** [comment]
released https://github.com/php-enqueue/enqueue-dev/blob/master/CHANGELOG.md#01021-2024-08-12

**Pull request merged**